File: hu.bme.mit.incqueryd.runtime/hu.bme.mit.incqueryd.core/src/main/java/hu/bme/mit/incqueryd/rete/actors/CoordinatorActor.java
package hu.bme.mit.incqueryd.rete.actors;

import static akka.pattern.Patterns.ask;
import hu.bme.mit.incqueryd.arch.ArchUtil;
import hu.bme.mit.incqueryd.rete.dataunits.Tuple;
import hu.bme.mit.incqueryd.rete.messages.CoordinatorCommand;
import hu.bme.mit.incqueryd.rete.messages.CoordinatorMessage;
import hu.bme.mit.incqueryd.rete.messages.Transformation;
import hu.bme.mit.incqueryd.rete.messages.YellowPages;
import hu.bme.mit.incqueryd.util.EObjectSerializer;
import hu.bme.mit.incqueryd.util.ReteNodeConfiguration;
import infrastructure.Cluster;
import infrastructure.Machine;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import org.eclipse.emf.common.util.EList;
import org.eclipse.emf.ecore.util.EcoreUtil;
import org.eclipse.incquery.runtime.rete.recipes.ProductionRecipe;
import org.eclipse.incquery.runtime.rete.recipes.ReteNodeRecipe;
import org.eclipse.incquery.runtime.rete.recipes.ReteRecipe;
import org.eclipse.incquery.runtime.rete.recipes.UniquenessEnforcerRecipe;

import scala.concurrent.Await;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;
import akka.actor.ActorRef;
import akka.actor.Address;
import akka.actor.Deploy;
import akka.actor.Props;
import akka.actor.UntypedActor;
import akka.remote.RemoteScope;
import akka.util.Timeout;
import arch.Configuration;
import arch.InfrastructureMapping;

public class CoordinatorActor extends UntypedActor {

	protected final boolean remoting;
	protected final String architectureFile;
	protected final Timeout timeout = new Timeout(Duration.create(14400, "seconds"));
	protected ActorRef productionActorRef;
	protected String query;
	protected boolean debug = false;
	protected Set<Tuple> latestResults;

	public CoordinatorActor(final String architectureFile, final boolean remoting) {
		super();
		this.architectureFile = architectureFile;
		this.remoting = remoting;

		if (architectureFile.contains("poslength")) {
			query = "PosLength";
		}
		if (architectureFile.contains("routesensor")) {
			query = "RouteSensor";
		}
		if (architectureFile.contains("signalneighbor")) {
			query = "SignalNeighbor";
		}
		if (architectureFile.contains("switchsensor")) {
			query = "SwitchSensor";
		}
	}

	public void start() throws Exception {
		final Configuration conf = ArchUtil.loadConfiguration(architectureFile);
		processConfiguration(conf);
	}

	// Rete recipe <-> IP address mapping
	final Map<ReteNodeRecipe, String> recipeToIp = new HashMap<>();
	// Rete recipe <-> ActorRef
	final Map<ReteNodeRecipe, ActorRef> recipeToActorRef = new HashMap<>();
	// EMF URI <-> Rete recipe
	final Map<String, ReteNodeRecipe> emfUriToRecipe = new HashMap<>();
	// EMF URI <-> ActorRef
	final Map<String, ActorRef> emfUriToActorRef = new HashMap<>();
	// collection of ActorRefs
	final Collection<ActorRef> actorRefs = new HashSet<>();

	private void processConfiguration(final Configuration conf) throws Exception {
		// mapping
		fillRecipeToIp(conf);

		// phase 1
		deployActors(conf);
		// create mapping based on the results of phase one mapping
		fillEmfUriToActorRef();

		// phase 2
		subscribeActors(conf);

		// phase 3
		initialize();
	}

	private void fillRecipeToIp(final Configuration conf) {
		final EList<InfrastructureMapping> mappings = conf.getMappings();
		for (final InfrastructureMapping mapping : mappings) {
			final Machine machine = mapping.getTargetElement();

			final EList<ReteNodeRecipe> mappedElements = mapping.getMappedElements();
			for (final ReteNodeRecipe reteNodeRecipe : mappedElements) {
				recipeToIp.put(reteNodeRecipe, machine.getIp());
			}
		}
	}

	private void fillEmfUriToActorRef() {
		for (final Entry<String, ReteNodeRecipe> emfUriAndRecipe : emfUriToRecipe.entrySet()) {
			final String emfUri = emfUriAndRecipe.getKey();
			final ReteNodeRecipe recipe = emfUriAndRecipe.getValue();

			final ActorRef akkaUri = recipeToActorRef.get(recipe);

			emfUriToActorRef.put(emfUri, akkaUri);

			if (debug)
				System.err.println("EMF URI: " + emfUri + ", Akka URI: " + akkaUri + ", traceInfo "
						+ ArchUtil.oneLiner(recipe.getTraceInfo()));
		}

		if (debug)
			System.err.println();
	}

	/**
	 * Phase 1
	 * 
	 * @param conf
	 * @throws Exception
	 */
	private void deployActors(final Configuration conf) throws Exception {
		final Cluster cluster = conf.getClusters().get(0);
		final EList<Machine> cacheMachines = cluster.getCacheMachines();
		
		final Collection<String> cacheMachineIps = new HashSet<>(); 
		for (final Machine machine : cacheMachines) {
			cacheMachineIps.add(machine.getIp());
		}
		
		for (final ReteRecipe rr : conf.getReteRecipes()) {
			for (final ReteNodeRecipe rnr : rr.getRecipeNodes()) {
				if (debug)
					System.err.println("[TestKit] Recipe: " + rnr.getClass().getName());

				final String ipAddress = recipeToIp.get(rnr);
				final String emfUri = EcoreUtil.getURI(rnr).toString();
				if (debug)
					System.err.println("[TestKit] - IP address:  " + ipAddress);
				if (debug)
					System.err.println("[TestKit] - EMF address: " + emfUri);
				emfUriToRecipe.put(emfUri, rnr);

				// create a clone, else we would get a java.util.ConcurrentModificationException
				final ReteNodeRecipe rnrClone = EcoreUtil.copy(rnr);
				final String recipeString = EObjectSerializer.serializeToString(rnrClone);

				final Props props;
				if (remoting) {
					props = new Props(ReteActor.class).withDeploy(new Deploy(new RemoteScope(new Address("akka",
							IncQueryDMicrokernel.ACTOR_SYSTEM_NAME, ipAddress, 2552))));
				} else {
					props = new Props(ReteActor.class);
				}
				final ActorRef actorRef = getContext().actorOf(props);

				configure(actorRef, recipeString, cacheMachineIps);

				actorRefs.add(actorRef);
				recipeToActorRef.put(rnr, actorRef);

				if (rnr instanceof ProductionRecipe) {
					productionActorRef = actorRef;
				}

				if (debug)
					System.err.println("[TestKit] Actor configured.");
				if (debug)
					System.err.println();
			}

		}
		if (debug)
			System.err.println("[ReteActor] All actors deployed and configured.");
		if (debug)
			System.err.println();
	}

	/**
	 * Phase 2
	 * 
	 * @param conf
	 * @throws Exception
	 */
	private void subscribeActors(final Configuration conf) throws Exception {
		final YellowPages yellowPages = new YellowPages(emfUriToActorRef);

		for (final ActorRef actorRef : actorRefs) {
			final Future<Object> future = ask(actorRef, yellowPages, timeout);
			Await.result(future, timeout.duration());
		}

		if (debug)
			System.err.println();
		if (debug)
			System.err.println();

		for (final Entry<String, ActorRef> entry : yellowPages.getEmfUriToActorRef().entrySet()) {
			if (debug)
				System.err.println(entry);
		}
	}

	/**
	 * Phase 3: initialize the Rete network.
	 * 
	 * @throws Exception
	 */
	private void initialize() throws Exception {
		final Collection<Future<Object>> futures = new HashSet<>();

		// send an INITIALIZE message to every "input actor"
		// in the current implementation input actors are described by a UniquenessEnforcerRecipe
		for (final Entry<ReteNodeRecipe, ActorRef> entry : recipeToActorRef.entrySet()) {
			final ReteNodeRecipe recipe = entry.getKey();
			if (recipe instanceof UniquenessEnforcerRecipe) {
				final ActorRef actorRef = entry.getValue();
				final Future<Object> future = ask(actorRef, CoordinatorMessage.INITIALIZE, timeout);
				futures.add(future);
			}
		}

		if (debug)
			System.err.println("<AWAIT>");
		for (final Future<Object> future : futures) {
			if (debug)
				System.err.println("await for " + future);
			final Object result = Await.result(future, timeout.duration());
			if (debug)
				System.err.println(result);
		}
		if (debug)
			System.err.println("</AWAIT>");
	}

	/**
	 * Phase 4: retrieve the results
	 * 
	 * @return
	 * 
	 * @throws Exception
	 */
	public Set<Tuple> check() throws Exception {
		latestResults = getQueryResults();
		if (debug)
			System.err.println("Results: " + latestResults.size());
		return latestResults;
	}

	/**
	 * Phase 5: do the transformation
	 * 
	 * @throws Exception
	 */
	public void transform() throws Exception {
		final long start = System.nanoTime();

		for (final Entry<ReteNodeRecipe, ActorRef> entry : recipeToActorRef.entrySet()) {
			final ReteNodeRecipe recipe = entry.getKey();
			if (recipe instanceof UniquenessEnforcerRecipe) {
				final UniquenessEnforcerRecipe uer = (UniquenessEnforcerRecipe) recipe;

				final ActorRef actorRef = entry.getValue();

				switch (query) {
				case "PosLength":
					if (uer.getTraceInfo().contains("Segment")) {
						final Transformation transformation = new Transformation(latestResults, query);
						final Future<Object> future = ask(actorRef, transformation, timeout);
						Await.result(future, timeout.duration());
					}
					break;
				case "RouteSensor":
					if (uer.getTraceInfo().contains("TrackElement_sensor")) {
						final Transformation transformation = new Transformation(latestResults, query);
						final Future<Object> future = ask(actorRef, transformation, timeout);
						Await.result(future, timeout.duration());
					}
					break;
				case "SignalNeighbor":
					if (uer.getTraceInfo().contains("Route_exit")) {
						final Transformation transformation = new Transformation(latestResults, query);
						final Future<Object> future = ask(actorRef, transformation, timeout);
						Await.result(future, timeout.duration());
					}
					break;
				case "SwitchSensor":
					if (uer.getTraceInfo().contains("TrackElement_sensor")) {
						final Transformation transformation = new Transformation(latestResults, query);
						final Future<Object> future = ask(actorRef, transformation, timeout);
						Await.result(future, timeout.duration());
					}
					break;
				}
			}
		}
	}

	private Set<Tuple> getQueryResults() throws Exception {
		final Future<Object> queryResultFuture = ask(productionActorRef, CoordinatorMessage.GETQUERYRESULTS, timeout);
		final Set<Tuple> result = (Set<Tuple>) Await.result(queryResultFuture, timeout.duration());
		return result;
	}

	private void configure(final ActorRef actorRef, final String recipeString, final Collection<String> cacheMachineIps) throws Exception {
		final ReteNodeConfiguration conf = new ReteNodeConfiguration(recipeString, cacheMachineIps);
		final Future<Object> future = ask(actorRef, conf, timeout);
		Await.result(future, timeout.duration());
	}

	@Override
	public void onReceive(final Object message) throws Exception {
		if (message == CoordinatorCommand.START) {
			start();
			getSender().tell(CoordinatorMessage.DONE, getSelf());
		} else if (message == CoordinatorCommand.CHECK) {
			final List<Tuple> results = new ArrayList<>(check());
			getSender().tell(results, getSelf());
		} else if (message == CoordinatorCommand.TRANSFORM) {
			transform();
			getSender().tell(CoordinatorMessage.DONE, getSelf());
		}
	}

}

File: hu.bme.mit.incqueryd.runtime/hu.bme.mit.incqueryd.core/src/main/scala/hu/bme/mit/incqueryd/rete/actors/CoordinatorActor.scala
package hu.bme.mit.incqueryd.rete.actors

import java.util.ArrayList
import java.util.HashMap
import java.util.HashSet
import java.util.Set
import scala.collection.JavaConversions._
import scala.concurrent.Await
import scala.concurrent.Future
import scala.concurrent.duration.Duration
import org.eclipse.emf.ecore.util.EcoreUtil
import org.eclipse.incquery.runtime.rete.recipes.BinaryInputRecipe
import org.eclipse.incquery.runtime.rete.recipes.InputRecipe
import org.eclipse.incquery.runtime.rete.recipes.ProductionRecipe
import org.eclipse.incquery.runtime.rete.recipes.ProjectionIndexerRecipe
import org.eclipse.incquery.runtime.rete.recipes.ReteNodeRecipe
import akka.actor.Actor
import akka.actor.ActorRef
import akka.actor.Address
import akka.actor.Deploy
import akka.actor.Props
import akka.pattern.Patterns.ask
import akka.remote.RemoteScope
import akka.util.Timeout
import arch.CacheRole
import arch.Configuration
import arch.ReteRole
import hu.bme.mit.incqueryd.arch.util.ArchUtil
import hu.bme.mit.incqueryd.monitoring.actors.JVMMonitoringActor
import hu.bme.mit.incqueryd.rete.dataunits.ChangeSet
import hu.bme.mit.incqueryd.rete.dataunits.ChangeType
import hu.bme.mit.incqueryd.rete.dataunits.ScalaChangeSet
import hu.bme.mit.incqueryd.rete.dataunits.Tuple
import hu.bme.mit.incqueryd.rete.messages.CoordinatorCommand
import hu.bme.mit.incqueryd.rete.messages.CoordinatorMessage
import hu.bme.mit.incqueryd.rete.messages.Transformation
import hu.bme.mit.incqueryd.rete.messages.YellowPages
import hu.bme.mit.incqueryd.retemonitoring.metrics.MonitoredActorCollection
import hu.bme.mit.incqueryd.retemonitoring.metrics.MonitoredMachines
import hu.bme.mit.incqueryd.retemonitoring.metrics.MonitoringMessage
import hu.bme.mit.incqueryd.util.EObjectSerializer
import hu.bme.mit.incqueryd.util.ReteNodeConfiguration
import infrastructure.Machine
import hu.bme.mit.bigmodel.fourstore.FourStoreLoader
import hu.bme.mit.bigmodel.fourstore.FourStoreDriver
import org.eclipse.incquery.runtime.rete.recipes.UnaryInputRecipe
import org.eclipse.incquery.runtime.rete.recipes.TypeInputRecipe
import infrastructure.Process
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import hu.bme.mit.incqueryd.rete.messages.UpdateMessage
import hu.bme.mit.incqueryd.rete.dataunits.ReteNodeSlot
import scala.collection.immutable.Stack
import hu.bme.mit.incqueryd.rete.messages.TerminationMessage
import java.nio.file.Paths
import org.apache.commons.io.FilenameUtils

class CoordinatorActor(val architectureFile: String, val remoting: Boolean) extends Actor {

  val logPrefix = "[CoordinatorActor] "
  val conf: Configuration = ArchUtil.loadConfiguration(architectureFile)
  val queryName = FilenameUtils.removeExtension(Paths.get(architectureFile).getFileName.toString())
  
  println("query name: " + queryName)
  
  var engine: ActorRef = null

  var verbose = true
  val timeout = new Timeout(Duration.create(14400, "seconds"))
  var productionActorRef: ActorRef = null
  var monitoringActor: ActorRef = null
  var yellowPages: YellowPages = null
  var latestResults = new HashSet[Tuple]
  var pendingUpdateMessages = 0

  var recipeToProcess = new HashMap[ReteNodeRecipe, Process]
  var recipeToActorRef = new HashMap[ReteNodeRecipe, ActorRef]
  var recipeToEmfUri = HashBiMap.create[ReteNodeRecipe, String]
  var emfUriToActorRef = new HashMap[String, ActorRef]
  var jvmActorRefs = new HashSet[ActorRef]
  
  def start = {
    processConfiguration
  }

  def processConfiguration = {
    // mapping
    fillRecipeToProcess

    // phase 1
    deployActors
    // deploy jvm monitoring actors as well
    deployJVMMonitoringActors

    // create mapping based on the results of phase one mapping
    fillEmfUriToActorRef

    // phase 2
    if (conf.getMonitoringMachine != null) {
      subscribeMonitoringService
    }

    subscribeActors

    // load the model to the Rete network
    load
  }

  def fillRecipeToProcess = {
    conf.getMappings.foreach(mapping => {
      mapping.getRoles.foreach(role => role match {
        case reteRole: ReteRole => recipeToProcess.put(reteRole.getNodeRecipe, mapping.getProcess)
      })
    })
  }

  def fillEmfUriToActorRef = {
    recipeToEmfUri.entrySet.foreach(emfUriAndRecipe => {
      val recipe = emfUriAndRecipe.getKey
      val emfUri = emfUriAndRecipe.getValue
      val akkaUri = recipeToActorRef.get(recipe)

      emfUriToActorRef.put(emfUri, akkaUri)
    })
  }

  def deployActors = {
    val cacheMachineIps = conf.getMappings.toList.
      filter(_.getRoles.exists(_.isInstanceOf[CacheRole])).
      map(_.getProcess.getMachine.getIp)

    if (verbose) println(logPrefix + "Deploying actors")
    conf.getRecipes.foreach(recipe =>
      recipe.getRecipeNodes.foreach(recipeNode => {
        if (verbose) println(logPrefix + "Recipe: " + recipeNode.getClass.getSimpleName)

        val emfUri = EcoreUtil.getURI(recipeNode).toString
        recipeToEmfUri.put(recipeNode, emfUri)

        // create a clone, else we would get a java.util.ConcurrentModificationException
        val rnrClone = EcoreUtil.copy(recipeNode)
        val recipeString = EObjectSerializer.serializeToString(rnrClone)

        var props: Props = null
        if (remoting) {
          val process = recipeToProcess.get(recipeNode)
          val ipAddress = process.getMachine.getIp
          val port = process.getPort

          if (verbose) println(logPrefix + "IP address:  " + ipAddress)
          if (verbose) println(logPrefix + "EMF address: " + emfUri)

          props = Props[ReteActor].withDeploy(new Deploy(new RemoteScope(new Address("akka",
            IncQueryDMicrokernel.ACTOR_SYSTEM_NAME, ipAddress, port))))
        } else {
          props = Props[ReteActor]
        }

        val actorRef = context.actorOf(props)

        configure(actorRef, recipeString, cacheMachineIps)

        recipeToActorRef.put(recipeNode, actorRef)

        recipeNode match {
          case pRec: ProductionRecipe => {
            if (pRec.getTraceInfo().contains(queryName)) {
              productionActorRef = actorRef
            }
          }
          case _ => {}
        }

        if (verbose) println(logPrefix + "Actor configured.")
      }))

    if (verbose) println(logPrefix + "All actors deployed and configured.")
    if (verbose) println

  }

  def deployJVMMonitoringActors = {
    if (remoting) {
      conf.getMappings.foreach(mapping => {
        val ipAddress = mapping.getProcess.getMachine.getIp
        val port = mapping.getProcess.getPort

        var props = Props[JVMMonitoringActor].withDeploy(new Deploy(new RemoteScope(new Address("akka",
          IncQueryDMicrokernel.ACTOR_SYSTEM_NAME, ipAddress, port))))

        val actorRef = context.actorOf(props)
        jvmActorRefs.add(actorRef)
      })
    } else {
      var props = Props[JVMMonitoringActor]
      val actorRef = context.actorOf(props)
      jvmActorRefs.add(actorRef)
    }

  }

  def subscribeActors = {
    if (verbose) println(logPrefix + "Subscribing actors")

    yellowPages = new YellowPages(emfUriToActorRef, monitoringActor)

    recipeToActorRef.foreach(recipeAndActorRef => {
      println(logPrefix + "Subscribing " + recipeAndActorRef._1 + " actor, which is based on a " + recipeAndActorRef + " recipe.")
      val future = ask(recipeAndActorRef._2, yellowPages, timeout)
      Await.result(future, timeout.duration)
    })
    
    if (verbose) println(logPrefix + "Actors subscribed.")
    if (verbose) println
  }

  def check = {
    val latestChangeSets = getQueryResults

    println(latestChangeSets.size + " lCS")
    println(latestChangeSets)
    
    latestChangeSets.foreach(latestChangeSet => {
      latestChangeSet.getChangeType match {
        case ChangeType.POSITIVE => latestResults.addAll(latestChangeSet.getTuples)
        case ChangeType.NEGATIVE => latestResults.removeAll(latestChangeSet.getTuples)
        case _ => {}
      }

      if (monitoringActor != null) monitoringActor ! sendChangesForMonitoring(latestChangeSet)
    })

    if (verbose) println("Results: " + latestResults.size)

    latestChangeSets
  }

  def sendChangesForMonitoring(changeSet: ChangeSet) = {
    val sb = new StringBuilder

    changeSet.getChangeType match {
      case ChangeType.POSITIVE => sb ++= "+ "
      case ChangeType.NEGATIVE => sb ++= "- "
    }

    changeSet.getTuples.foreach(tuple => {
      for (i <- 0 to tuple.size - 1) {
        sb ++= tuple.get(i) + ":"
      }
      sb.deleteCharAt(sb.size - 1)
      sb += ';'
    })

    sb.toString
  }

  def load = {
    println(logPrefix + "Loading the Rete network...")

    val clusterName = conf.getConnectionString.split("://")(1)
    val databaseDriver = new FourStoreDriver(clusterName)

    conf.getRecipes.foreach(recipe =>
      recipe.getRecipeNodes.foreach(_ match {
        case typeInputRecipe: TypeInputRecipe =>

          val tuples = scala.collection.mutable.Set[Tuple]()
          typeInputRecipe match {
            case binaryInputRecipe: BinaryInputRecipe => { 
              val traceInfo = binaryInputRecipe.getTraceInfo
              if (traceInfo.startsWith("attribute")) {
                initializeAttribute(databaseDriver, binaryInputRecipe, tuples)
              } else if (traceInfo.startsWith("edge")) {
                  initializeEdge(databaseDriver, binaryInputRecipe, tuples)
              }
            }
            case unaryInputRecipe: UnaryInputRecipe => {
              initializeVertex(databaseDriver, unaryInputRecipe, tuples)
            }
          }

          val actor = recipeToActorRef.get(typeInputRecipe)
          val changeSet = new ChangeSet(tuples, ChangeType.POSITIVE)

          update(actor, changeSet)
        case _ => {}
      }))
  }

  /**
   * Update API
   */
  def update(actor: ActorRef, changeSet: ChangeSet) {
    val senderStack = Stack(self)
    val updateMessage = new UpdateMessage(changeSet, ReteNodeSlot.SINGLE, senderStack)

    // send the updates to the actor
    println(logPrefix + "Sending update message to " + actor)
    actor.tell(updateMessage)

    pendingUpdateMessages += 1
    println(logPrefix + pendingUpdateMessages + " update message(s) pending.")
  }

  def initializeAttribute(databaseDriver: FourStoreDriver, recipe: BinaryInputRecipe, tuples: scala.collection.mutable.Set[Tuple]) = {
    val attributes = databaseDriver.collectVerticesWithProperty(recipe.getTypeName)

    attributes.foreach(attribute => {
      tuples += new Tuple(attribute._1, attribute._2)
    })
  }

  def initializeEdge(databaseDriver: FourStoreDriver, recipe: BinaryInputRecipe, tuples: scala.collection.mutable.Set[Tuple]) = {
    val edges = databaseDriver.collectEdges(recipe.getTypeName)

    edges.entries.foreach(edge => {
      tuples += new Tuple(edge.getKey, edge.getValue)
    })
  }

  def initializeVertex(databaseDriver: FourStoreDriver, recipe: UnaryInputRecipe, tuples: scala.collection.mutable.Set[Tuple]) = {
    val vertices = databaseDriver.collectVertices(recipe.getTypeName)
    vertices.foreach(vertex => tuples += new Tuple(vertex))
  }

  def getQueryResults: java.util.List[ChangeSet] = {
    val queryResultFuture = ask(productionActorRef, CoordinatorMessage.GETQUERYRESULTS, timeout)
    Await.result(queryResultFuture, timeout.duration).asInstanceOf[java.util.List[ChangeSet]]
  }

  def configure(actorRef: ActorRef, recipeString: String, cacheMachineIps: List[String]) = {
    val reteConf = new ReteNodeConfiguration(recipeString, cacheMachineIps, conf.getConnectionString)
    val future = ask(actorRef, reteConf, timeout)
    Await.result(future, timeout.duration)
  }

  def subscribeMonitoringService = {
    monitoringActor = context.actorFor("akka://monitoringserver@" + conf.getMonitoringMachine.getIp + ":5225/user/collector")
    monitoringActor ! new MonitoredActorCollection(recipeToActorRef.values, jvmActorRefs)
  }

  def receive = {
    case CoordinatorCommand.START => {
      start
      engine = sender
    }
    case CoordinatorCommand.CHECK => {
      sender ! check
    }
    case CoordinatorCommand.TRANSFORM => {
      sender ! CoordinatorMessage.DONE
    }
    case CoordinatorCommand.LOAD => {
      load
      sender ! CoordinatorMessage.DONE
    }
    case terminationMessage: TerminationMessage => {
      pendingUpdateMessages -= 1
      println(logPrefix + "Termination message received, " + pendingUpdateMessages + " update message(s) pending.")

      if (pendingUpdateMessages == 0) {
        println(logPrefix + "Termination protocol finished.")
        engine ! CoordinatorMessage.DONE
      }
    }
    case _ => {}
  }

}